proxy/src/main/java/com/velocitypowered/proxy/plugin/loader/JavaPluginLoader.java

```diff
@@ -0,0 +1,129 @@
+
+import com.google.inject.Guice;
+import com.google.inject.Injector;
+import com.velocitypowered.api.plugin.*;
+import com.velocitypowered.api.plugin.meta.PluginDependency;
+import com.velocitypowered.api.proxy.ProxyServer;
+import com.velocitypowered.proxy.VelocityServer;
+import com.velocitypowered.proxy.plugin.PluginClassLoader;
+import com.velocitypowered.proxy.plugin.loader.java.JavaVelocityPluginDescription;
+import com.velocitypowered.proxy.plugin.loader.java.SerializedPluginDescription;
+import com.velocitypowered.proxy.plugin.loader.java.VelocityPluginModule;
+
+import javax.annotation.Nonnull;
+import java.io.BufferedInputStream;
+import java.io.InputStreamReader;
+import java.io.Reader;
+import java.net.URL;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.HashSet;
+import java.util.Optional;
+import java.util.Set;
+import java.util.jar.JarEntry;
+import java.util.jar.JarInputStream;
+import java.util.regex.Pattern;
+
+public class JavaPluginLoader implements PluginLoader {
+    private final ProxyServer server;
+    private final Path baseDirectory;
+
+    public JavaPluginLoader(ProxyServer server, Path baseDirectory) {
+        this.server = server;
+        this.baseDirectory = baseDirectory;
+    }
+
+    @Nonnull
+    @Override
+    public PluginDescription loadPlugin(Path source) throws Exception {
+        Optional<SerializedPluginDescription> serialized = getSerializedPluginInfo(source);
+
+        if (!serialized.isPresent()) {
+            throw new InvalidPluginException("Did not find a valid velocity-info.json.");
+        }
+
+        PluginClassLoader loader = new PluginClassLoader(
+                new URL[] {source.toUri().toURL() }
+        );
+
+        Class mainClass = loader.loadClass(serialized.get().getMain());
+        VelocityPluginDescription description = createDescription(serialized.get(), source, mainClass);
+
+        String pluginId = description.getId();
+        Pattern pattern = PluginDescription.ID_PATTERN;
+
+        if (!pattern.matcher(pluginId).matches()) {
+            throw new InvalidPluginException("Plugin ID '" + pluginId + "' must match pattern " + pattern.pattern());
+        }
+
+        return description;
+    }
+
+    @Nonnull
+    @Override
+    public PluginContainer createPlugin(PluginDescription description) throws Exception {
+        if (!(description instanceof JavaVelocityPluginDescription)) {
+            throw new IllegalArgumentException("Description provided isn't of the Java plugin loader");
+        }
+
+        JavaVelocityPluginDescription javaDescription = (JavaVelocityPluginDescription) description;
+        Optional<Path> source = javaDescription.getSource();
+
+        if (!source.isPresent()) {
+            throw new IllegalArgumentException("No path in plugin description");
+        }
+
+        Injector injector = Guice.createInjector(new VelocityPluginModule(server, javaDescription, baseDirectory));
```

Comment:
Allow plugin to implement their module for in-plugin injections.

Reply:
I don't know how I think about this. This is an area of active consideration.

Reply:
Maybe special field in `Plugin` annotation that holds list of modules.(`Class<Module>[]`). Every module instance created using some injector, then plugin injector created with these modules.